pkg/controllers/allocation/controller.go

```diff
@@ -66,7 +66,16 @@ func NewController(kubeClient client.Client, coreV1Client corev1.CoreV1Interface
 // Reconcile executes an allocation control loop for the resource
 func (c *Controller) Reconcile(ctx context.Context, object client.Object) (reconcile.Result, error) {
 	provisioner := object.(*v1alpha2.Provisioner)
-	// 1. Filter pods
+
+	// 1. Complement provisioner with (dynamic) default values, which must not
+	//    be hoisted into the original CRD as they might change with each reconciliation
+	//    loop iteration.
+	provisionerWithDefaults, err := provisioner.WithDynamicDefaults()
```

Comment:
Consider naming this `provisioner` and naming the other one `persisted`, so the rest of the function doesn't need to be aware that anything has been defaulted.

Reply:
Actually, my motivation was to make sure the rest of the functions is aware of this and does not by accident persist (apply/update/patch/...) the `provisionerWithDefaults` object. Maybe a better approach to make intentions more clear and to prevent other code from accidentally persisting this `transient` object is to return `ProvisionerSpec` (instead of `Provisioner`) from `WithDynamicDefaults()` and to pass that spec to all the cloud provider methods.

Reply:
We actually used to have this, but found it was useful to be able to use the provisioner name/namespace in caching logic (though I don't believe we do this anymore). You could also imagine that cloud providers may wish to customize behavior (beyond labels) by adding an annotation to the metadata on the provisioner. Perhaps it's better to discourage this though.

In the interest of gaining the value the PR brings already, I'm happy to punt on this question and leave as is. I'm also supportive of moving the cloud provider API to deal only with ProvisionerSpec to make the mutation question more clear.

Reply:
OK, I will keep that `ProvisionerSpec` refactoring out of this PR to keep is simple and focused. I guess that would first need a discussion/design on how to pass cloud provider specific configuration as you already mused about in https://github.com/awslabs/karpenter/issues/501#issuecomment-873117461.